### ISSUE TITLE
better handling of flags

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#892]
Issue #829 made steps toward dealing with flagged data in `argo` objects, but as stated in the discussion there it is worth thinking about how to do more than just set all flagged data to NA. For example, with argo data, I might want to set all `flag==3` and `flag==4` data to NA, but not 1 or 2.

One way to do this was suggested in #829, was to provide a `key` argument, like:
```r
handleFlags(x, key=list(bad=c(4, NA), good=1)) # keep only data known to be good
handleFlags(x, key=list(bad=c(4, NA), good=c(0,1))) # also keep the un-QC'd data
```

Another way would be for `handleFlags()` to have an arg for flag type and replacement, like:
```r
handleFlags(x, flag=4, replace=NA) # replace 4 with NA
handleFlags(x, flag=c(3, 4, NA), replace=NA) # replace flag 3, 4 and NA with NA
```
I suppose it's possible that the `replace` arg could be a function, though it's not obvious to me if that would be useful or not.

On a related point, for an `argo` object currently the flags are saved as characters in a matrix. Does it make sense for that to be converted to numeric at read time, for easier comparison and plotting?

**Comment by dankelley** [comment]
To be honest, I've lost track of what  was meant by `handle_flags`. I guess it does the extractions, in which case I could maybe argue that `subset` would be more easily guessed.

Getting the interface right is important because it's no fun for users when things change. So I think we should toss various ideas back and forth, and -- importantly -- we should type in the code that users would be writing, to see what "feel" right.

I envisage something like
```r
plot(subset(x, CONDITION))
```
or maybe
```r
plot(handleFlags(x, ARGUMENTS))
```
and the first feels better to me because it's in the R convention.

I sort of thought that the purpose of `handle_flags` might be to get around the fact that flags are different for different data types. Thus, maybe
```r
new <- subset(handleFlags(x), CONDITION)
```
would be the idea. Then `CONDITION` could be the same for different `class(x)`. For example
```r
plotTS(subset(handleFlags(x), salinityFlag==0&temperatureFlag==0&pressureFlag=0)
```
would do something good whether `x` was a `ctd`, a `section`, an `argo`, etc.

This doesn't "feel" too bad to me.  Of course, as noted, I completely forget what `handle_flags` was thought to be doing. (I had a quick look at #829 but couldn't quite tell.) So maybe the idea would be 
```r
plotTS(handleFlags(x))
```
and the idea is that `handleFlags` *removes* data.  My only issue with this, compared to the more complex notation I'm giving above, is that `handleFlags` won't know *why* the user is handling flags. For example, do we need good S, T and p, or is it okay to just have T? (I have worked with data where the S was bad but I still wanted the T, and if the p was good, I'd use a TS diagram to at least get something, and that would be quite good if we just had heaving.)  The `subset` way makes it clear.

I'll be in CDOGS for most of the day and won't likely look at emails.

**Comment by richardsc** [comment]
I agree this is a bit complicated ... the sort of conversation that would benefit from a coffee and standing at a blackboard. I'm not in a rush for this -- I do want to get it right, though.

I'm just writing, but here's an "off the top of my head" list for things I'd like our "handle flags" function to be:

* flexible across classes -- e.g. perhaps even a generic?
* ability to remove or replace flagged values according to a defined key. That key could either be provided in the function call, or even added to the object `@metadata` (e.g. for `argo` objects the key is fixed)
* updates the processing log accordingly
* sensible defaults, but configurable to be able to deal with cases like you mentioned -- when you want to toss S, but keep T and p (even if flagged)

Going to keep thinking about this ... might make sense to have a real conversation about it at some point if we think we're not making headway here.

**Comment by dankelley** [comment]
I agree it should be a new generic. I've never made new generics before but I know it's very easy. The only slight snag is documenting them with roxygen2. I don't think any new code should be documented with the old-style manually created files in `man/`. I'm sure it's easy to document, and that stackexchange has good advice.

**Important.** I was about to type that the work could be done in the `flags` branch, but I just did a trial merge and there are conflicts so I just didn't bother with it ... I may look tonight. I find it difficult to keep track of old branches, generally. This is why I like to merge back into develop quickly, and also why I always press to have issue titles be narrow-grained so that the branches are only active for a short time.

This brings up another point -- I think I should merge `vector_imu` into `develop`, so we won't run into merge conflicts a month or two from now, when (if) nortek gives fuller documentation on the imu format. I wonder if you can have a look at the question I posed at #893 on this. I think what `oce` does right now on imu is the right thing (It tells people what is in the `.vec` file) so I would like to merge the branch and then remove it so I don't get mixed up later. OK?

I have a meeting at 10AM my time but if you wanted to chat on the phone other times, maybe just send an email 1/2 hour before so I'll be at my desk.

**Comment by dankelley** [comment]
I decided I was annoyed with myself that the `flags` branch was so old, so I merged `develop` into it, fixed a bunch of conflicts, pushed, checked the diff to `develop` to see that things looked either good or at least not obviously bad, and then popped back to `develop` and merged `flags` into it.

That's a long way of stating that (a) `flags` is now up to date with `develop` and (b) `develop` might have just got some new bugs.  I know (b) is bad, but I figured there might be something in `flags` that was useful (and my `git diff` suggested there were some nice things there).

I feel a phonecall might be a good way to figure out who does what on this and, more importantly, exactly what will be done. I'll be back from my class at 1130 my time tomorrow, if you wanted to try then. (This paragraph will get deleted from the web after I comment ... that way, you see it in the email but other people don't see one-to-one information.)

**Comment by dankelley** [comment]
I just did some tests, and am putting here as a reminder. I don't know how to document yet.
```r
library(oce)
setGeneric(name="handleFlags", def=function(object){standardGeneric("handleFlags")})
data(ctd)
setMethod(f="handleFlags",signature="oce",definition=function(object) {return(1.23)})
handleFlags(ctd)
```
So, this works. Tonight I'll figure out how to document with Roxygen2. But I will only be pushing to the `kelley` branch since I am not assigned to the issue.

**Comment by dankelley** [comment]
In branch `kelley` I have put a preliminary test of generic `handleFlags`, called `NEWhandleFlags`, which has skeletons for both `ctd` and `adp`. Those skeletons are in `R/AllClass.R`, to permit easy cut/paste operations for new skeletons.  I need to do some more work on a method for documenting because in the end we'll have something like 15 variants and a lot of text should be the same in each. (Roxygen has a way to do that, I think.)

This stuff is not ready to be deployed, i.e. there is little point in trying to actually code in some functionality until the documentation scheme is settled. One thing I learned just in typing in the skeleton code is that our present funciton is too limited. It has just two args, one for the object and one for the action. I think `action` should be akin to a verb (e.g. the default, `"NA"`, means "change into NA"). But we need another argument, perhaps called `flags` (that's what I called it just for something to type in a rush, but it "feels" right), and probably in the form of a list or something. I think it's critical to think through the arg list because changing things later will be a pain.

**Comment by dankelley** [comment]
I had some time before my class and mocked-up a method for generics in the `kelley` branch (commit c70e757d7a3719a46ab9619f202ae5c490935f8d). I didn't bother coding the actual action. Instead, I learned how to make generics and -- harder, actually -- how to document them. I could explain how this works if needed. Probably some evening I will code in something for the CTD case -- the BIO people are keen on that.

**Comment by dankelley** [comment]
In `kelley` I have made a stab at documenting the new generic. Perhaps CR can try building this and doing
```R
?`NEWhandleFlags,ctd-method`
```
and sending comments on how it reads. It is a bit strained in spots because (by intention) some of this will be identical for all variants of the function.  I'm also interested in whether the ctd-specific parts read well, and look to be a good UI.

PS. the ctd docs use templates that are in `man-roxygen`. I'm hoping we can get the ctd docs looking clean and informative, before proceeding to code.

**Comment by dankelley** [comment]
It occurs to me that we should have ways to say "except". I've not written that in the docs but it could be e.g. `flags=list(temperature=-2)` to mean everything *but* 2 is bad (conservative WHP) or maybe `flags=list(temperature=c(-2,-3))` to also accept flag=3, which is "questionable" for WHP. I'll write that into the docs when I get some time today.

**Comment by dankelley** [comment]
Um, further to that last point, we cannot actually use negative for logical negation because there are flag codes of 0 being use. So, although I've pushed to `kelley` some text that says we use negatives, we cannot do that. Another syntax will be needed. (This proves the importance of writing docs first ... we could have 200 lines of code by now, doing stuff in a way that will break when someone tries to negate a zero code.) Maybe Clark can pull `kelley`, build, and do `?"NEWhandleFlags,ctd-method"` to see what I've written ... he will probably see how to denote logical negation.

**Comment by dankelley** [comment]
For reference, the ODV system has a list of flags, and there are certainly a *lot* of them!
https://odv.awi.de/fileadmin/user_upload/odv/misc/ODV4_QualityFlagSets.pdf

**Comment by dankelley** [comment]
I talked about inverse flags with CL today. One possibility would be an arg named, well, maybe that, or `flagsInverted`.  So e.g. 
```r
new <- handleFlags(old, flagsInverted=list(temperature=0))
```
would mean to set to `NA` all temperatures *except* when flag=0.  (Since `action` is not supplied above, the default of setting to `NA` will be performed; see https://github.com/dankelley/oce/blob/kelley/man-roxygen/handleFlagsTemplate.R for docs)

**Comment by dankelley** [comment]
This GH interface is weird; only one person can be assigned.

**Comment by richardsc** [comment]
Just off the top of my head, I wonder if the "inverted" behaviour we're going for can be built into the `action` argument, instead of making a new one. E.g. `action` doesn't have to be specific to a replacement value, but could maybe recognize a few "verbs" for what to do. Roughly, using the WHP flag example:
```r
## keep only good flagged temperature
ctdGood <- handleFlags(ctd, flags=list(temperature=2), actions=list(temperature="keep''))

## the other way
ctdGood <- handleFlags(ctd, flags=list(temperature=(1:10)[-2]), actions=list(temperature=NA))
```
both of these would accomplish the same thing. Which is better?

**Comment by dankelley** [comment]
Oh, very clever with the `1:10`. One thing I prefer about that is the issue of what to do with things that do *not* meet a criterion. For example
```r
handleFlags(ctd, flags=list(temperature=4), action=list(temperature="na"))
```
means to set certain temperatures to `NA`, but to *leave the others alone*.  The thing is, leaving alone is the same as "keep".  So, to me, the first of your styles does not quite feel natural. Of course, all it would require would be a phrase to *make* it feel natural.

I vote that we accept both methods you've listed in the [previous comment](https://github.com/dankelley/oce/issues/892#issuecomment-200948984).

**Comment by dankelley** [comment]
On the weekend, I revamped `handleFlags`. I also did a ton of changes to docs, because I have decided that the "family" roxygen2 tag is so useful that all the docs should be done in roxygen2.  (Another advantage is discoverability -- it names things in a consistent way, which can help users when they learn that way.)  I did all of this in the `kelley` branch, which I just merged into `develop`.

Clark - it would be great if you could have a look, to see how it "feels". I have not built in negation, because I think your `(0:100)[-2]` idea is a good solution to that problem. So far, the permitted actions are somewhat limited. I did not do `"keep"` because I think that is implied as the no-action scenario. I don't think it makes sense to explain the functioning here -- the docs should tell that. I didn't make a `flags_ctd` branch because the work was mainly general and because I was not making the office-home transition and so didn't need to push to a branch for continuity.

PS. the docs are using templates. I think I'll be doing that for other things, too, e.g #902

**Comment by richardsc** [comment]
I'm having a look at this now, for the `argo` class case. So far I think the syntax is pretty good, though this is a big enough change that I think we'll want to try it out in real work for sometime before we commit fully. I'm going to do this work on my own fork, until I think it's ready for a merge.

One thing that I noticed is that currently `read.argo()` creates `flags` fields with names like:
```r
> names(argo[['flags']])
[1] "pressureQc"            "pressureAdjustedQc"   
[3] "salinityQc"            "salinityAdjustedQc"   
[5] "temperatureQc"         "temperatureAdjustedQc"
```
which doesn't work with the current version of `handleFlagsInternal()`. I'm going to open a new issue (and assign myself), to fix this at the same time as I work on the `handleFlags,argo-method` stuff.

**Comment by dankelley** [comment]
Clark, is this ready to close? I'm finding it confusing with so many open and inter-related issues.  (This is why I've closed several issues in the past minute.)

**Comment by richardsc** [comment]
Yes, I think it's probably safe to close. I did notice that you closed #895, though I'm not sure that we've fully tested the implementation given the outstanding difference between `argo` and `ctd` behaviour reported here:

https://github.com/dankelley/oce/issues/896#issuecomment-205944491

Probably best to leave anything that comes up to it's own issue, though.
